lib/mazes/BeattieSchoberth.js

```diff
@@ -0,0 +1,785 @@
+        const width = opt('roomWidth');
+        const height = opt('roomHeight');
+        let lines = [];
+        for(let y=0; y < height; y++){
+            let row = [];
+            for(let x=0; x < width; x++){
+                if(x === 0 || x === width-1){
+                    //TODO: doors
+                    if(x=== 0){
+                        if(options.doors.w && inDoorRange(y, height, 2)){
+                            row.push(TILE.DOOR);
+                        }else{
+                            row.push(TILE.WALL);
+                        }
+                    }else{
+                        if(options.doors.e && inDoorRange(y, height, 2)){
+                            row.push(TILE.DOOR);
+                        }else{
+                            row.push(TILE.WALL);
+                        }
+                    }
+                }else{
+                    if(y === 0 || y === height-1){
+                        if(y=== 0){
+                            if(options.doors.n && inDoorRange(x, width, 2)){
+                                row.push(TILE.DOOR);
+                            }else{
+                                row.push(TILE.WALL);
+                            }
+                        }else{
+                            if(options.doors.s && inDoorRange(x, width, 2)){
+                                row.push(TILE.DOOR);
+                            }else{
+                                row.push(TILE.WALL);
+                            }
+                        }
+                    }else{
+                        row.push(TILE.FLOOR);
+                    }
+                }
+            }
+            lines.push(row);
+        }
+        return lines;
+    }
+    
+    emptyRoom(options={}){
+        const opt = (name) => options[name] || this.options[name];
+        const width = opt('roomWidth');
+        const height = opt('roomHeight');
+        let result = [];
+        for(let y=0; y < height; y++){
+            let row = [];
+            for(let x=0; x < width; x++){
+                row.push(TILE.VOID);
+            }
+            result.push(row);
+        }
+        return result;
+    }
+    
+    build(rand){
+        const adventureBuild = this.adventure.build(rand);
+        adventureBuild.world = this.renderWorld({}, adventureBuild);
+        return adventureBuild;
+    }
+    
+}
+
+export default function ALGORITHM_BEATTIE_SCHOBERTH(tileMap, options) {
+    let fractional = Math.floor(Math.sqrt(Math.max(tileMap.width, tileMap.height)));
+    
+    const builder = new BeattieSchoberthBuilder({
+        rooms: fractional*2, 
+        keys: 3,
+        special: 2,
+        width: tileMap.width,
+        height: tileMap.height,
+        roomHeight: options.roomHeight || fractional,
+        roomWidth: options.roomWidth || fractional
+    });
+    
+    const built = builder.build(()=>{
+        return tileMap.mersenneTwister.rand();
```

Comment:
Probably best to make calls directly into `tileMap.random()` instead of calling into `mersenneTwiter`. This will allow `TileClass` to delegate randomness to provider.